### PR TITLE
Grasp using proximity

### PR DESCRIPTION
**Pull request by 708yamaguchi**
use proximity sensor in :start-grasp

### CODE REVIEW

File: jsk_arc2017_baxter/euslisp/lib/arc-interface.l [hunk 1]
```diff
@@ -260,13 +260,16 @@
...
     (let (graspingp avs object-index obj-pos obj-cube pinch-yaw)
       ;; TODO: object-index is set randomly
       (setq object-index (random (length (gethash arm object-boxes-))))
       (setq obj-pos
             (send self :get-object-position arm movable-region :object-index object-index))
       (setq obj-cube (send self :bbox->cube (elt (gethash arm object-boxes-) object-index)))
-      (if (> (x-of-cube obj-cube) (y-of-cube obj-cube)) (setq pinch-yaw pi/2) (setq pinch-yaw 0))
+      (if (> (x-of-cube obj-cube) (y-of-cube obj-cube))
+          (setq pinch-yaw (+ pi/2 (caar (send (send obj-cube :worldcoords) :rpy-angle))))
+        (setq pinch-yaw (caar (send (send obj-cube :worldcoords) :rpy-angle))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/arc-interface.l]: Indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 2]
```diff
@@ -153,6 +164,9 @@
   (:set-finger-flex
     (arm side msg)
     (sethash side (gethash arm finger-flex-) (send msg :data)))
+  (:set-proximities
+   (arm msg)
+   (sethash arm proximities- (send msg :proximities)))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 3]
```diff
@@ -196,6 +214,22 @@
   (:get-prismatic-vel (arm)
     (send self :spin-once)
     (gethash arm prismatic-vel-))
+  (:get-proximity (arm side &key (raw nil))
+     (send self :spin-once)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 4]
```diff
@@ -196,6 +214,22 @@
...
+     (let ((rl (gethash :left (gethash arm proximity-init-values-)))
+           (rr (gethash :right (gethash arm proximity-init-values-)))
+           ;;(rm (gethash :middle (gethash arm proximity-init-values-)))
+           (proximities (gethash arm proximities-)))
+       (if (null raw)
+           (cond
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 5]
```diff
@@ -392,7 +463,18 @@
                (setq rarm-pressure-threshold- (- min-pressure 15))))))
     (send self :stop-grasp arm)
     (ros::ros-info "[:calib-pressure-threshold] Threshold r: ~a l: ~a"
-                   rarm-pressure-threshold- larm-pressure-threshold-)))
+                   rarm-pressure-threshold- larm-pressure-threshold-))
+  (:calib-proximity-threshold
+   (&optional (arm :arms))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 6]
```diff
@@ -225,9 +259,46 @@
...
+               ;; rotate arm using proximity sensor
                (setf (aref finger-av 1) 180)
-               (send self :move-hand l/r finger-av 1000)
+               (send self :move-hand l/r finger-av 2000 :wait nil)
+               (unix:usleep 10000)
+               (while (and
+                       (< (max (send self :get-proximity l/r :right)
+                               (send self :get-proximity l/r :left))
+                          proximity-threshold-)
+                       (setq hand-interpolatingp (send self :hand-interpolatingp))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: Indent

File: jsk_arc2017_baxter/euslisp/lib/baxter.l [hunk 7]
```diff
@@ -109,10 +109,11 @@
     (case (car args)
       ((:angle-vector)
        (let ((av (cadr args)) (joints (gethash arm hand-joints-)))
-         (when av
-           (dotimes (i (length av))
-             (send self (elt joints i) :joint-angle (elt av i))
-             (send self (elt joints (+ i (length av))) :joint-angle (elt av i))))
+         (if (and (null (eq (length av) 0)) (null (eq (length av) 2)))
+             (error "length of angle-vector must be 0 or 2.~%"))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter.l]: indent

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 8]
```diff
@@ -225,9 +259,46 @@
...
+                     (progn
+                        (send robot l/r :wrist-r :joint-angle 45 :relative t)
+                        (setq av (send robot :angle-vector))
+                        (send robot :angle-vector prev-av)
+                        (send self :angle-vector-raw av 2000)
+                        (send self :move-hand l/r finger-av 4000 :wait nil)
+                        (unix::usleep 10000)
+                        (while (and (< (send self :get-proximity l/r :right)
+                                       (send self :get-proximity l/r :left))
+                                    (send self :interpolatingp))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: Please add unix::usleep in this loop to reduce CPU load

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 9]
```diff
@@ -225,9 +259,46 @@
...
+                   (progn
+                     (send robot l/r :wrist-r :joint-angle -45 :relative t)
+                     (setq av (send robot :angle-vector))
+                     (send robot :angle-vector prev-av)
+                     (send self :angle-vector-raw av 2000)
+                     (send self :move-hand l/r finger-av 4000 :wait nil)
+                     (unix::usleep 10000)
+                     (while (and (> (send self :get-proximity l/r :right)
+                                    (send self :get-proximity l/r :left))
+                                 (send self :interpolatingp)))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: Please add (unix::usleep 1000) in this loop to reduce CPU load

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l (hunk 8 above)
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: I found `:wait-interpolation-until` also forgot sleeping
https://github.com/start-jsk/jsk_apc/blob/master/jsk_arc2017_baxter/euslisp/lib/baxter-interface.l#L375-L378
I'll fix this in another PR.

**Comment by pazeshun** [comment]
Nice! Close #2047

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 10]
```diff
@@ -225,9 +259,46 @@
...
+                        (while (and (< (send self :get-proximity l/r :right)
+                                       (send self :get-proximity l/r :left))
+                                    (send self :interpolatingp))))
+                   (progn
+                     (send robot l/r :wrist-r :joint-angle -45 :relative t)
+                     (setq av (send robot :angle-vector))
+                     (send robot :angle-vector prev-av)
+                     (send self :angle-vector-raw av 2000)
+                     (send self :move-hand l/r finger-av 4000 :wait nil)
+                     (unix::usleep 1000)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: You should not reduce this loop.
Add sleep in while loop below.

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 11]
```diff
@@ -225,9 +259,46 @@
...
+                 (send self :update-robot-state :wait-until-update t)
+                 (setq prev-av (send robot :angle-vector))
+                 (if (< (send self :get-proximity l/r :right) (send self :get-proximity l/r :left))
+                     (progn
+                        (send robot l/r :wrist-r :joint-angle 45 :relative t)
+                        (setq av (send robot :angle-vector))
+                        (send robot :angle-vector prev-av)
+                        (send self :angle-vector-raw av 2000)
+                        (send self :move-hand l/r finger-av 4000 :wait nil)
+                        (unix::usleep 1000)
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: You should not reduce this loop.
Add sleep in while loop below.

Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: Indent, 3 spaces -> 2 spaces after `progn`

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 12]
```diff
@@ -196,6 +214,22 @@
...
+      (if (null raw)
+        (cond
+         ;; for sparkfun proximity sensor
+         ((eq side :left)
+          (/ (- (send (elt proximities 0) :average) rl) (expt (/ rl 2500.0) 1.5)))
+         ((eq side :right)
+          (/ (- (send (elt proximities 1) :average) rr) (expt (/ rr 2500.0) 1.5))))
+        (cond
+         ((eq side :left) (send (elt proximities 0) :average))
+         ((eq side :right) (send (elt proximities 1) :average))))))
```
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: indent, 2 spaces after `cond`

File: jsk_arc2017_baxter/euslisp/lib/baxter.l [hunk 13]
```diff
@@ -109,10 +109,11 @@
     (case (car args)
       ((:angle-vector)
        (let ((av (cadr args)) (joints (gethash arm hand-joints-)))
-         (when av
-           (dotimes (i (length av))
-             (send self (elt joints i) :joint-angle (elt av i))
-             (send self (elt joints (+ i (length av))) :joint-angle (elt av i))))
+         (if (and (null (eq (length av) 0)) (null (eq (length av) 2)))
+           (error "length of angle-vector must be 0 or 2.~%"))
```
Comment (knorth55) [inline jsk_arc2017_baxter/euslisp/lib/baxter.l]: use `ros::ros-error` instead of `error`.

Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter.l]: I think this should be error.
If ros::ros-error, motion continues and may cause fatal result such as finger collision.

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 14]
```diff
@@ -225,9 +259,52 @@
...
+                     (setq av (send robot :angle-vector))
+                     (send robot :angle-vector prev-av)
+                     (send self :angle-vector-raw av 2000)
+                     (send self :move-hand l/r finger-av 4000 :wait nil)
+                     (dotimes (x 100)
+                       (if (send self :interpolatingp) (return))
+                       (unix::usleep 1000))
+                     (while (and (< (send self :get-proximity l/r :right)
+                                    (send self :get-proximity l/r :left))
+                                 (send self :interpolatingp))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: ごめんなさい、伝わってないみたいですが、このwhile文の中にsleepをいれてほしいです
この書き方だと、CPU使用率が100%になると思います

File: jsk_arc2017_baxter/euslisp/lib/baxter-interface.l [hunk 15]
```diff
@@ -225,9 +259,52 @@
...
+                     (setq av (send robot :angle-vector))
+                     (send robot :angle-vector prev-av)
+                     (send self :angle-vector-raw av 2000)
+                     (send self :move-hand l/r finger-av 4000 :wait nil)
+                     (dotimes (x 100)
+                       (if (send self :interpolatingp) (return))
+                       (unix::usleep 1000))
+                     (while (and (> (send self :get-proximity l/r :right)
+                                    (send self :get-proximity l/r :left))
+                                 (send self :interpolatingp)))))
```
Comment (pazeshun) [inline jsk_arc2017_baxter/euslisp/lib/baxter-interface.l]: ごめんなさい、伝わってないみたいですが、このwhile文の中にsleepをいれてほしいです
この書き方だと、CPU使用率が100%になると思います

**Pull request merged**